diyepw/get_noaa_isd_lite_file.py

```diff
@@ -0,0 +1,118 @@
+import pkg_resources
+from urllib.error import URLError
+
+from ._logging import _logger
+from .exceptions import DownloadNotAllowedError
+
+def get_noaa_isd_lite_file(wmo_index:int, year:int, *, output_dir:str = None, allow_downloads:bool = False) -> str:
+    """
+    Given a WMO index and a year, retrieve the corresponding NOAA ISD Lite AMY file
+    :param wmo_index:
+    :param year:
+    :param output_dir: Optional output directory - if not specified, the file will be saved to a package directory.
+        If the directory already contains a NOAA ISD Lite file matching the requested WMO Index and year, then a new
+        file will not be downloaded from NOAA and that file's path will be returned
+    :param allow_downloads: Pass True to permit NOAA ISD Lite files and related information to be downloaded from
+        ncdc.noaa.gov if they are not already present in output_dir.
+    :return: The path to the NOAA ISD Lite file
+    """
+    if output_dir is None: # pragma: no cover
+        output_dir = pkg_resources.resource_filename('diyepw', 'data/noaa_isd_lite_files')
+        _logger.debug(f"get_noaa_isd_lite_file() - output_dir was not defined, will use {output_dir}")
+
+    if not os.path.exists(output_dir): # pragma: no cover
+        os.mkdir(output_dir)
+        _logger.debug(f"get_noaa_isd_lite_file() - {output_dir} did not exist, so has been created")
+
+    # On the NOAA website, the ISD Lite files are named with a third number between WMO and year, but
+    # since we don't use that third number for anything and it complicates identifying a file for a
+    # WMO/Year combination, we simplify the name to only contain the values we care about
+    file_name = f"{wmo_index}-{year}.gz"
+    file_path = os.path.join(output_dir, file_name)
+
+    # Download the ISD Lite file if it's not already in the output directory
+    if not os.path.exists(file_path):
+        url = _get_noaa_isd_lite_file_url(year, wmo_index, allow_downloads)
+
+        if not allow_downloads:
+            raise DownloadNotAllowedError(
+                f"The ISD Lite file {file_path} is not present. Pass allow_downloads=True to allow the "
+                f"missing data to be automatically downloaded from {url}"
+            )
+
+        try:
+            with request.urlopen(url) as response:
+                with open(file_path, 'wb') as downloaded_file:
+                    downloaded_file.write(response.read())
+        except URLError as e: # pragma: no cover
+            raise Exception(f'Failed to download {url} - are you connected to the internet?')
+        except Exception as e: # pragma: no cover
+            raise Exception(f"Error downloading from {url}: {e}")
+
+    return file_path
+
+def _get_noaa_isd_lite_file_url(year:int, wmo_index:int, allow_downloads:bool) -> str:
+    catalog = _get_noaa_isd_lite_file_catalog(year, allow_downloads=allow_downloads)
+    file_name = list(catalog.loc[catalog['wmo_index'] == wmo_index]['file_name'])[0]
```

Comment:
I'm having some trouble with this line:
- first, it fails if the wmo_index i've chosen doesn't exist
- second, `list` constructor doesn't seem to like taking in the Series... but series has a `.tolist()` method or you could use `.iloc[0]`

Reply:
I wonder if we have different Python or library versions? I'm running 3.8.5.

I'll switch to using .iloc, though, that's the more idiomatic way to do it. This was one of the earlier chunks I wrote and I wasn't very familiar with manipulating data with Pandas, so a lot of this stuff is just the one way I figured out that worked at the time :)

Reply:
I've also added a proper exception with a useful error message if get_noaa_isd_lite_file is passed an invalid WMO index.